### PR TITLE
Ignore subset aggregates

### PR DESCRIPTION
**Pull request by terencechain**
Implements and rationale: https://github.com/ethereum/consensus-specs/pull/2847
 
This PR drops aggregated objects if a better aggregate has been seen. (i.e. bitfield overlaps). As a result, we should see a reduction in CPU usage and incoming and outgoing bandwidth. 

Note: we did not have to do anything with aggregated attestation, `HasAggregatedAttestation` already checks the subset 🙌🏼

### CODE REVIEW

File: beacon-chain/sync/service.go [hunk 1]
```diff
@@ -221,6 +223,7 @@ func (s *Service) initCaches() {
 	s.seenUnAggregatedAttestationCache = lruwrpr.New(seenUnaggregatedAttSize)
 	s.seenSyncMessageCache = lruwrpr.New(seenSyncMsgSize)
 	s.seenSyncContributionCache = lruwrpr.New(seenSyncContributionSize)
+	s.syncContributionBitsOverlapCache = lruwrpr.New(seenSyncContributionSize)
```
Comment (terencechain) [inline beacon-chain/sync/service.go:226]: For simplicity, I opted in same cache size as `seenSyncContributionSize`, open for other suggestions

File: beacon-chain/sync/validate_sync_contribution_proof.go [hunk 2]
```diff
@@ -318,6 +340,55 @@ func (s *Service) setSyncContributionIndexSlotSeen(slot types.Slot, aggregatorIn
...
+	v, ok := s.syncContributionBitsOverlapCache.Get(string(b))
+	if !ok {
+		s.syncContributionBitsOverlapCache.Add(string(b), [][]byte{c.AggregationBits.Bytes()})
+		return nil
+	}
+	bitsList, ok := v.([][]byte)
+	if !ok {
+		return errors.New("could not covert cached value to []bitfield.Bitvector")
+	}
+	s.syncContributionBitsOverlapCache.Add(string(b), append(bitsList, c.AggregationBits.Bytes()))
```
Comment (nisdas) [inline beacon-chain/sync/validate_sync_contribution_proof.go:358]: you need to check whether it contains a bitlist that already contains over what the contribution has.

File: beacon-chain/sync/validate_sync_contribution_proof.go [hunk 3]
```diff
@@ -148,6 +152,24 @@ func rejectEmptyContribution(m *ethpb.SignedContributionAndProof) validationFn {
...
+	return func(ctx context.Context) (pubsub.ValidationResult, error) {
+		c := m.Message.Contribution
+		seen, err := s.hasSeenSyncContributionBits(c)
+		if err != nil {
+			return pubsub.ValidationIgnore, err
+		}
+		if seen {
+			return pubsub.ValidationIgnore, nil
+		}
+		seen = s.hasSeenSyncContributionIndexSlot(c.Slot, m.Message.AggregatorIndex, types.CommitteeIndex(c.SubcommitteeIndex))
```
[the remark this answers is not in the document]
Comment (terencechain) [reply beacon-chain/sync/validate_sync_contribution_proof.go:164]: thanks, I removed `ignoreCached`. I could also use `ignoreCached` here too if you prefer that

File: beacon-chain/sync/validate_sync_contribution_proof.go (hunk 2 above)
Comment (terencechain) [reply beacon-chain/sync/validate_sync_contribution_proof.go:355]: fixed, thanks!

File: beacon-chain/sync/validate_sync_contribution_proof.go [hunk 4]
```diff
@@ -318,6 +330,67 @@ func (s *Service) setSyncContributionIndexSlotSeen(slot types.Slot, aggregatorIn
 	s.seenSyncContributionCache.Add(string(b), true)
 }
 
+// Set sync contribution's slot, root, committee index and bits.
+func (s *Service) setSyncContributionBits(c *ethpb.SyncCommitteeContribution) error {
+	s.syncContributionBitsOverlapLock.Lock()
+	defer s.syncContributionBitsOverlapLock.Unlock()
+	b := append(c.BlockRoot, bytesutil.Bytes32(uint64(c.Slot))...)
```
Comment (nisdas) [inline beacon-chain/sync/validate_sync_contribution_proof.go:337]: I think you should copy `c.BlockRoot` first before appending it here. Due to how protobuf unmarshalling is carried out, you could unintentionally mutate this sync contribution.

File: beacon-chain/sync/validate_sync_contribution_proof.go [hunk 5]
```diff
@@ -318,6 +330,67 @@ func (s *Service) setSyncContributionIndexSlotSeen(slot types.Slot, aggregatorIn
...
+	}
+	bitsList, ok := v.([][]byte)
+	if !ok {
+		return false, errors.New("could not covert cached value to []bitfield.Bitvector128")
+	}
+	return BitListOverlaps(bitsList, c.AggregationBits.Bytes())
+}
+
+// BitListOverlaps returns true if there's an overlap between two bitlists.
+func BitListOverlaps(bitLists [][]byte, b []byte) (bool, error) {
```
Comment (nisdas) [inline beacon-chain/sync/validate_sync_contribution_proof.go:377]: no need to export it here, as no consumer is going to use this from the sync package currently.

File: proto/prysm/v1alpha1/sync_committee_mainnet.go [hunk 6]
```diff
@@ -10,3 +10,7 @@ import (
 func NewSyncCommitteeAggregationBits() bitfield.Bitvector128 {
 	return bitfield.NewBitvector128()
 }
+
+func ConvertSyncContributionBitVector(b []byte) bitfield.Bitvector128 {
```
Comment (nisdas) [inline proto/prysm/v1alpha1/sync_committee_mainnet.go:14]:
```suggestion
func ConvertToSyncContributionBitVector(b []byte) bitfield.Bitvector128 {
```

File: proto/prysm/v1alpha1/sync_committee_minimal.go [hunk 7]
```diff
@@ -10,3 +10,7 @@ import (
 func NewSyncCommitteeAggregationBits() bitfield.Bitvector8 {
 	return bitfield.NewBitvector8()
 }
+
+func ConvertSyncContributionBitVector(b []byte) bitfield.Bitvector8 {
```
Comment (nisdas) [inline proto/prysm/v1alpha1/sync_committee_minimal.go:14]:
```suggestion
func ConvertToSyncContributionBitVector(b []byte) bitfield.Bitvector8 {
```

**Review by nisdas — APPROVED** [review]

**Review by nisdas — APPROVED** [review]

**Pull request merged**